code/modules/library/skill_learning/skill_station.dm

```diff
@@ -132,51 +145,84 @@
-	chip.on_apply(target, silent = FALSE)
-	chip.forceMove(target_brain)
-	LAZYADD(target_brain.skillchips, chip)
+	to_chat(carbon_occupant, "<span class='notice'>Operation complete!</span>")
+
+	SStgui.update_uis(src)
 
 /obj/machinery/skill_station/ui_data(mob/user)
 	. = ..()
 	.["working"] = working
 	.["timeleft"] = work_timer ? timeleft(work_timer) : null
 	var/mob/living/carbon/carbon_occupant = occupant
-	var/obj/item/organ/brain/occupant_brain = carbon_occupant.getorganslot(ORGAN_SLOT_BRAIN)
-	if(QDELETED(carbon_occupant) || QDELETED(occupant_brain))
-		.["error"] = "Brain not detected. Please consult nearest medical practitioner."
-	else
-		var/list/current_skills = list()
-		for(var/obj/item/skillchip/skill_chip in occupant_brain)
-			current_skills += list(list("name"=skill_chip.skill_name,"icon"=skill_chip.skill_icon,"cost"=skill_chip.slot_cost,"ref"=REF(skill_chip),"active"=(skill_chip in occupant_brain.skillchips)))
-		.["current"] = current_skills
-		.["slots_used"] = carbon_occupant.used_skillchip_slots
-		.["slots_max"] = carbon_occupant.max_skillchip_slots
 
 	.["skillchip_ready"] = inserted_skillchip ? TRUE : FALSE
 	if(inserted_skillchip)
-		.["implantable"] = inserted_skillchip.can_be_implanted(occupant)
-		.["implantable_reason"] = inserted_skillchip.can_be_implanted_message(occupant)
+		// This is safe, incompatibility check can accept a null or invalid mob.
+		var/incompatibility_check = inserted_skillchip.has_mob_incompatibility(carbon_occupant)
+		if(incompatibility_check)
+			.["implantable"] = FALSE
+			.["implantable_reason"] = incompatibility_check
+		else
+			.["implantable"] = TRUE
+			.["implantable_reason"] = null
 		.["skill_name"] = inserted_skillchip.skill_name
 		.["skill_desc"] = inserted_skillchip.skill_description
 		.["skill_icon"] = inserted_skillchip.skill_icon
 		.["skill_cost"] = inserted_skillchip.slot_cost
 
+	// If there's no occupant, we don't need to worry about what skillchips are in their brain.
+	if(!carbon_occupant)
+		.["error"] = "No valid occupant detected. Please consult nearest medical practitioner."
+		.["current"] = null
+		.["slots_used"] = null
+		.["slots_max"] = null
+		return
+
+	var/obj/item/organ/brain/occupant_brain = carbon_occupant?.getorganslot(ORGAN_SLOT_BRAIN)
+
+	// If there's no brain, we don't need to worry about what skillchips are in it but can still process
+	// any info about the skillslot capacity of the body, even though the info is useless without a brain.
+	if(QDELETED(occupant_brain))
+		.["error"] = "Brain not detected. Please consult nearest medical practitioner."
+		.["current"] = null
+		.["slots_used"] = carbon_occupant.used_skillchip_slots
+		.["slots_max"] = carbon_occupant.max_skillchip_slots
+		return
+
+	// If we got here, we have both an occupant and it has a brain, so we can check for skillchips.
+	var/list/current_skills = list()
+	for(var/obj/item/skillchip/skill_chip in occupant_brain.skillchips)
```

Comment:
Previously this was all skillchips located in the brain because you want to see/remove inactive skillchips.

Reply:
All skillchips should be located in brain.skillchips now, regardless of whether they are active or not.

There's been code refactoring around this area.

Skillchip activity is now denoted by a simple TRUE/FALSE flag that is set when the skillchip is activated or deactivated. In essence, it tells you if the last operation called was on_activate or on_deactivate and as such tells you whether the body has the effects of the skillchip or not.

Activating and de-activating are now events seperate to implating and removing the skillchip, as a result.

There should be no need to check outside of brain.skillchips - All skillchips currently jammed into the brain should be in that list.

Naturally /mob/living/carbon/proc/update_skillchips() is updated as a result. And has had more descriptive var names added so Rohesie doesn't lynch me.

Expect to see that with the review PR, which will likely trigger another review due to additional code changes. Although I'm hoping I've squashed all the stuff Rohesie keeps yelling at me for.